engine/src/flutter/impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -50,496 +52,581 @@ class PositionWriter {
-                 /*reverse=*/false);
-        continue;
-      }
+  void RecordQuad(Point p1, Point cp, Point p2) override {
+    RecordCurve<PathTessellator::Quad>({p1, cp, p2});
+  }
 
-      previous_offset = offset;
-      offset = ComputeOffset(contour_start_point_i, contour_start_point_i,
-                             contour_end_point_i, contour);
-      const Point contour_first_offset = offset.GetVector();
-
-      // Generate start cap.
-      if (!polyline.contours[contour_i].is_closed) {
-        Point cap_offset =
-            Vector2(-contour.start_direction.y, contour.start_direction.x) *
-            stroke_width * 0.5f;  // Counterclockwise normal
-        cap_proc(vtx_builder, polyline.GetPoint(contour_start_point_i),
-                 cap_offset, scale, /*reverse=*/true);
-      }
+  void RecordConic(Point p1, Point cp, Point p2, Scalar weight) override {
+    RecordCurve<PathTessellator::Conic>({p1, cp, p2, weight});
+  }
 
-      for (size_t contour_component_i = 0;
-           contour_component_i < contour.components.size();
-           contour_component_i++) {
-        const Path::PolylineContour::Component& component =
-            contour.components[contour_component_i];
-        bool is_last_component =
-            contour_component_i == contour.components.size() - 1;
-
-        size_t component_start_index = component.component_start_index;
-        size_t component_end_index =
-            is_last_component ? contour_end_point_i - 1
-                              : contour.components[contour_component_i + 1]
-                                    .component_start_index;
-        if (component.is_curve) {
-          AddVerticesForCurveComponent(
-              vtx_builder, component_start_index, component_end_index,
-              contour_start_point_i, contour_end_point_i, contour);
-        } else {
-          AddVerticesForLinearComponent(
-              vtx_builder, component_start_index, component_end_index,
-              contour_start_point_i, contour_end_point_i, contour);
+  void RecordCubic(Point p1, Point cp1, Point cp2, Point p2) override {
+    RecordCurve<PathTessellator::Cubic>({p1, cp1, cp2, p2});
+  }
+
+  template <typename Curve>
+  inline void RecordCurve(const Curve& curve) {
+    auto start_direction = curve.GetStartDirection();
+    auto end_direction = curve.GetEndDirection();
+
+    // The Prune receiver should have eliminated any empty curves, so any
+    // curve we see should have both start and end direction.
+    FML_DCHECK(start_direction.has_value() && end_direction.has_value());
+
+    // In order to keep the compiler/lint happy we check for values anyway.
+    if (start_direction.has_value() && end_direction.has_value()) {
+      // We now know the curve cannot be degenerate
+      auto start_perpendicular = ComputePerpendicular(-start_direction.value());
+      auto end_perpendicular = ComputePerpendicular(end_direction.value());
```

Comment:
Remove `auto`s here and below.

Reply:
That's not a valid exception =)

https://google.github.io/styleguide/cppguide.html#Type_deduction

Reply:
Done